no-vary-query.md

```diff
@@ -0,0 +1,264 @@
+- [Detailed design](#detailed-design)
+  - [The header](#the-header)
+  - [Integration with…](#integration-with)
+    - [… HTTP caches](#-http-caches)
+    - [… preloading caches](#-preloading-caches)
+    - [… the cache API](#-the-cache-api)
+    - [… other browser caches](#-other-browser-caches)
+  - [Interaction with prerendering activation](#interaction-with-prerendering-activation)
+  - [Interaction with redirects](#interaction-with-redirects)
+  - [Interaction with storage partitioning](#interaction-with-storage-partitioning)
+- [Alternatives considered](#alternatives-considered)
+- [Future extensions](#future-extensions)
+  - [More complex no-vary rules](#more-complex-no-vary-rules)
+  - [`No-Vary-Path`](#no-vary-path)
+  - [A referrer hint](#a-referrer-hint)
+- [Security and privacy considerations](#security-and-privacy-considerations)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Goals
+
+- Allow caches to avoid keying on the order of URL query parameters
+
+- Allow caches to avoid keying on specifically-named URL query parameters
+
+- Allow caches to avoid keying on all URL query parameters, or all-but-some-specifically-named URL query parameters
+
+## Non-goals
+
+- Allow more complicated keying rules than the above, e.g. "treat `key` and `KEY` as equivalent"; or, "treat `value1` and `value2` as equivalent for `key`, but treat `value3` differently"; or, "order doesn't matter for most parameters, but the position of the `key` parameter matters"; or, "allow the `width` parameter to match as long as its value is within 100 of the already-cached URL's `width` value". Some of these [seem possible as future extensions](#more-complex-no-vary-rules), but we are not designing around them.
+
+- Support non-standard query parameter structures, e.g. `?color=red;size=large` with a `;` delimiter instead of a `&`. We will use the [`application/x-www-form-urlencoded` format](https://url.spec.whatwg.org/#concept-urlencoded) supported by the URL Standard.
+
+- Allow caches to avoid keying on other parts of the URL, notably the path. Although this could be useful, we think it would best be done via [an additive separate solution](#no-vary-path).
+
+## Prior art
+
+This proposal takes some of its inspiration from the existing [`Vary`](https://httpwg.org/specs/rfc7231.html#header.vary) HTTP header, which lets servers indicate what _header names_ should be _included_ when constructing the cache key. Unlike for headers, by default responses vary on their URL (including query string), so our proposal uses the `No-Vary-` prefix to show how it's indicating which query parameters should be _excluded_ when constructing the cache key. We see these two headers, the existing `Vary` and our proposed `No-Vary-Query`, as part of a general cache-key-construction mechanism, and want all places on the web platform that respect the former to also respect the latter.
```

Comment:
The "all places on the web platform..." statement is quite strong. If we want to make it, it would be good to also note that any software which doesn't implement `No-Vary-Query` wouldn't behave incorrectly, it would just have fewer cache hits than it might otherwise.